### PR TITLE
Allow for not broadcasting

### PR DESCRIPTION
**Pull request by chriselrod**

### CODE REVIEW

**Comment by chriselrod** [comment]
@YingboMa `@.. broadcast=false` is basically what the DiffEq ecosystem always wants, so I think we should merge this and maybe update the diffeq ecosytem.

We could also consider making `broadcast=false` the default, bumping the version to `0.2` instead, which'd make updating the DiffEq ecosystem easier (bumping compat instead of changing actual uses).

`@broadcast=false` disables broadcasting, meaning it assumes all dynamic axes are of the same size.
This removes the check and fallback codegen.

File: src/FastBroadcast.jl [hunk 1]
```diff
@@ -263,27 +263,46 @@ function broadcast_codeinfo(ci, threadarg)
...
+    fb_macro(ex, __module__, threadarg, broadcastarg)
+end
+macro (..)(kwarg0, kwarg1, ex)
+    threadarg, broadcastarg = _process_kwarg(kwarg0)
+    threadarg, broadcastarg = _process_kwarg(kwarg1, threadarg, broadcastarg)
+    fb_macro(ex, __module__, threadarg, broadcastarg)
+end
+
+@generated function fast_materialize!(dst, ::DB, bc::Broadcasted, dstaxes::Tuple{Vararg{Any,N}}, ax, indexstyle) where {N,DB}
+    broadcast = DB === True
```
Comment (YingboMa) [inline src/FastBroadcast.jl:305]: Is this variable ever used?

File: test/runtests.jl [hunk 2]
```diff
@@ -16,13 +16,22 @@ if GROUP == "All" || GROUP == "Core"
...
         @test (@.. (x*y) + x + y + x + y + x + y) == bcref
         @test (@.. thread=true dest .+= (x*y) + x + y + x + y + x + y) ≈ 2bcref
         @test (@.. dest .-= (x*y) + x + y + x + y + x + y) ≈ bcref
         @test (@.. thread=true dest *= (x*y) + x + y + x + y + x + y) ≈ abs2.(bcref)
+
+        @test (@.. broadcast=false dest = (x*y) + x + y + x + y + x + y) == bcref
+        @test (@.. broadcast=false (x*y) + x + y + x + y + x + y) == bcref
+        @test (@.. broadcast=false thread=true dest .+= (x*y) + x + y + x + y + x + y) ≈ 2bcref
+        @test (@.. broadcast=false dest .-= (x*y) + x + y + x + y + x + y) ≈ bcref
+        @test (@.. broadcast=false thread=true dest *= (x*y) + x + y + x + y + x + y) ≈ abs2.(bcref)
```
Comment (YingboMa) [inline test/runtests.jl:32]: What if there's a broadcast when `broadcast=false`? Should we throw a readable error?

Comment (chriselrod) [reply test/runtests.jl:32]: Your call. What did the old `@..` do?

File: src/FastBroadcast.jl (hunk 1 above)
Comment (chriselrod) [reply src/FastBroadcast.jl]: Oops.

File: test/runtests.jl (hunk 2 above)
Comment (YingboMa) [reply test/runtests.jl:32]: It gave a really ugly error message :p so maybe we can do better. Could be a different PR though.

**Comment by YingboMa** [comment]
> We could also consider making broadcast=false the default, bumping the version to 0.2 instead, which'd make updating the DiffEq ecosystem easier (bumping compat instead of changing actual uses).

Yeah, I agree. Could you also update README.md and add a note on the different behavior?

File: src/FastBroadcast.jl [hunk 3]
```diff
@@ -310,20 +328,25 @@ end
...
+    if DB === False
+        push!(q.args, loop_quote)
+    else
+        push!(q.args, :(if isfast
             $loop_quote
         else
             Base.Cartesian.@nloops $N i dst begin
                 $loopbody_slow
             end
-        end
```
Comment (chriselrod) [inline src/FastBroadcast.jl:324]: Basically, if `DB === False`, we'd just have to replace
```julia
            Base.Cartesian.@nloops $N i dst begin
                $loopbody_slow
            end
```
with
```julia
throw($(helpful error message))
```

**Comment by YingboMa** [comment]
BTW, is Static 0.5 OK? https://github.com/YingboMa/FastBroadcast.jl/blob/ab5234ea294405e05870ecef2339b9594cb4d140/Project.toml#L13

**Comment by chriselrod** [comment]
> BTW, is Static 0.5 OK?
 
Probably?

**Review by YingboMa — APPROVED** [review]

**Pull request merged**